open source intra_op_parallel (#19175)

jspark1105 opened the pull request:
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/19175

Open source operators with intra-op parallelism

TODO: gate tbb dependent code

Differential Revision: D14417823

JianpingChen066 commented:
looks like this patch doesn't include any CMakeLists.txt to build the *.cc under intra_op_parallel into  libcaffe2 ?  May the build modification can be include in this patch also thus I can have some try the test on parallel_mlp_benchmark.py ?

JianpingChen066 commented:
Thanks for CMakeLists.txt added. 
I have tried and meet two issues here:
1) If I just use "cmake -DUSE_TBB=ON ../" to build, it will report error like:
../tbb_numa_all_reduce_op.cc:192:5: error: ‘_mm_sfence’ was not declared in this scope
     _mm_sfence();
     ^~~~~~~~~~
for tbb_numa_all_reduce_op.cc, I added "#include <immintrin.h>" to make compilation pass.
2) When I try to run the parallel_mlp_benchmark.py as below
  export PYTHONPATH=$HOME/pytorch/build_tbb/
  python ./parallel_mlp_benchmark.py --num-workers 1 --num-numa-nodes 2
It will report the error as:
    libcaffe2.so: undefined symbol: _ZTIN3tbb4taskE
Should CMakeLists.txt also need to add something like
  +if (USE_TBB)
  +       target_link_libraries(caffe2 PUBLIC tbb)
  +       target_link_libraries(caffe2 PUBLIC rt)
  +endif()

Thanks

jspark1105 commented:
Thanks @JianpingChen066 I'll apply your suggestions!

JianpingChen066 commented:
3) And in parallel_mlp_benchmark.py, 
"from caffe2.python.intra_op_parallel.parallel_benchmark_utils" should be "caffe2.intra_op_parallel.parallel_benchmark_utils" ? 

Regards

jspark1105 committed:
open source intra_op_parallel (#20121)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20121

Pull Request resolved: https://github.com/pytorch/pytorch/pull/19175

Open source operators with intra-op parallelism

TODO: gate tbb dependent code

Reviewed By: dmudiger

Differential Revision: D14417823

fbshipit-source-id: 5e88d414df60a45447cdaab39ad6686d1bd5ae54

jgong5 commented on caffe2/intra_op_parallel/intra_op_parallel.cc:
```diff
@@ -0,0 +1,199 @@
+        FLAGS_caffe2_intra_op_parallel_only_grab_idle_threads && !use_tbb_
+            ? pool->numAvailable() + 1
+            : pool->size());
+  }
```
So it means there would be oversubscription issue with TBB?

anton-malakhov replied:
It's the number of tasks/chunks, not the threads. TBB actually like when there are more tasks than CPUs, so it can balance load better. Oversubscription is bad with threads, which are managed by the TBB in a process-wide thread pool and reused between task arenas and algorithms.

jgong5 commented on caffe2/intra_op_parallel/intra_op_parallel.cc:
```diff
@@ -0,0 +1,199 @@
+        FLAGS_caffe2_intra_op_parallel_only_grab_idle_threads && !use_tbb_
+            ? pool->numAvailable() + 1
+            : pool->size());
+  }
```
So it means there would be oversubscription issue with TBB?

jgong5 commented on caffe2/intra_op_parallel/tbb_numa_all_reduce_op.h:
```diff
@@ -0,0 +1,63 @@
+  typedef ::tbb::flow::continue_node<::tbb::flow::continue_msg> cn_type;
+
+  std::vector<std::unique_ptr<::tbb::flow::graph>> dags_;
+  // TODO: reuse arena in net_tbb_task_graph
```
How about having a TBB-specific threading pool like `TBBThreadPool` to wrap the task arena object, as a counterpart of `TaskThreadPool`?

jspark1105 replied:
Thanks for the suggestion! Definitely sounds a good idea. Let us take a stab at this.

JianpingChen066 commented on caffe2/intra_op_parallel/parallel_benchmark_utils.py:
```diff
@@ -0,0 +1,218 @@
+            "MergeDim",
+            [_get_sharded_blob_name(blob_name, numa_node_id)],
+            _get_sharded_blob_name(blob_name, numa_node_id),
+            device_option=default_device_option,
```
I just wonder why "MergeDim" for all numa-nodes use "device_option=default_device_option" and not use "device_option=caffe2_pb2.DeviceOption(numa_node_id=numa_node_id)" like "FC" ops ? 
Thanks

skywalker-2015 commented on CMakeLists.txt:
```diff
@@ -123,6 +123,7 @@ option(USE_ROCKSDB "Use RocksDB" OFF)
 option(USE_SNPE "Use Qualcomm's SNPE library" OFF)
 option(USE_SYSTEM_EIGEN_INSTALL
     "Use system Eigen instead of the one under third_party" OFF)
+option(USE_TBB "Use Thread Builing Block library" OFF)
```
```suggestion
option(USE_TBB "Use Threading Building Blocks library" OFF)
```
